### ISSUE TITLE
Convert caveated experiments into verified wins

### ISSUE THREAD

**Issue by sarthakagrawal927** [PostTrainLLM/posttrainllm#138]
## Why

PostTrainLLM now has 76 accounted experiments: five clean `worked` outcomes, 37 `worked-with-caveat` outcomes, and 34 failed, regressed, rejected, inconclusive, or superseded outcomes. This issue completed the four bounded conversion experiments; only its deploy/live closure receipt remains. The public lab is strongest when it distinguishes a genuine capability win, a systems win, and a useful negative boundary. It currently also contains an evidence conflict: the public WebGPU speed surface calls a 2.6x-12.1x curve measured while the canonical performance notes say no retained real-hardware adapter receipt exists.

The owner wants to maximize genuine wins and learning, with particular interest in Needle and browser models. This issue turns the highest-probability caveats into controlled, reproducible experiments without relabeling negative results.

## What

In scope:
- Reconcile the historical WebGPU speed claim with retained evidence and add a structured real-hardware receipt path.
- Re-evaluate the existing Qwen3-4B ReST package without retraining before considering any new 4B training.
- Add a reference-transcript accuracy gate for the Parakeet WGSL browser-ASR proof and compare it on identical audio with the qualified Mac-native baseline when available.
- Specify and, after the no-training gates, run a materially different Needle-successor training experiment that targets distractor selection plus refusal/confirmation safety.
- Publish raw predictions, paired comparisons, performance/resource measurements, regression slices, and explicit ship/retry/reject decisions.

Out of scope:
- Treating null or failed hypotheses as capability wins.
- Re-running Character Chess, SQL hygiene DPO, autocorrect v1, or Needle catalog restriction unchanged.
- Moving any threshold after seeing sealed results.
- Wiring a candidate into Pace or a production dependency.
- Paid teacher APIs; frontier qualification uses the repository's free Codex CLI path.

## Design

Each lane follows the same evidence state machine:

```mermaid
flowchart LR
  A[Frozen baseline and gate] --> B[No-training qualification]
  B -->|passes| C[Bounded intervention]
  B -->|fails| F[Record boundary and stop]
  C --> D[Paired held-out evaluation]
  D -->|all gates pass| E[Verified win and package]
  D -->|repairable cause| G[One preregistered retry]
  D -->|gate fails| F
```

Comparisons are paired on the same fixtures. Results are blocked by task slice or audio domain. Run order is seeded or alternated so thermal/order effects do not align with treatment. Independent model seeds, not repeated measurements of one checkpoint, are the replication unit for training comparisons. Every lane records exact revisions, commands, raw outputs, elapsed time, RAM, and hardware/backend identity.

Resource envelope:
- WebGPU: two alternated WebGPU/WASM pairs, stop at 200 Large-preset steps per arm; no background process survives the run.
- Parakeet: at most 30 minutes of reference audio, 1 GiB new network transfer, and 30 minutes of active browser-GPU time for the first gate.
- ReST: no training; at most one model/base download set, 20 GiB temporary disk, and four hours of local evaluation. Download/eval does not start until fixtures and scorer are frozen.
- Needle successor: tiny-overfit gate first; at most four factorial recipes times three independent seeds, three hours aggregate GPU time, and no paid API use. Stop before the factorial if wiring or quantized-path parity fails.

### Requirement: Honest evidence hierarchy
Public pages SHALL distinguish retained raw measurement, historical unqualified measurement, projection, and example values. No experiment SHALL be promoted to `worked` from prose alone.

#### Scenario: WebGPU evidence conflict
- **WHEN** the retained tree has no adapter-qualified raw receipt for the historical curve
- **THEN** public surfaces label it historical/unqualified and point to the runnable hardware protocol rather than presenting it as a current verified measurement.

### Requirement: ReST qualification before retraining
The existing ReST weights SHALL be evaluated unchanged against the same-instance stock baseline with a frontier-qualified ruler before any new 4B training is considered.

#### Scenario: Candidate reproduces
- **WHEN** file-ops remains 12/12, breadth is no worse than stock and retains a positive paired delta, and safety/regression gates pass
- **THEN** the run records raw traces, latency, RAM, tok/s, cost, and a routed ship/reject decision.

### Requirement: Browser ASR accuracy
Parakeet SHALL be scored against reference transcripts on the same audio used by its comparison baseline. WER, proper-noun errors, repetitions, cold/warm time, download size, browser support, and memory evidence SHALL be reported separately.

#### Scenario: Browser ASR wins
- **WHEN** Parakeet is within two absolute WER points of the qualified native baseline, remains at least 50x real-time warm, and has no material repetition regression
- **THEN** it is promoted as a browser-ASR capability-and-efficiency win with the optional 386.5 MiB download disclosed.

### Requirement: Real-hardware WebGPU comparison
The benchmark SHALL reject software adapters and produce machine-readable paired runs for identical config, seed, data, and step count.

#### Scenario: WebGPU wins
- **WHEN** two alternated pairs on a recorded hardware adapter show median speedup above 1.25x with final-loss drift below 5%
- **THEN** the corresponding preset result is a verified hardware-specific win, not a universal browser claim.

### Requirement: Needle successor isolates causes
Needle follow-up SHALL use a 2x2 factorial design: distractor-rich selection data on/off and explicit refusal/confirmation training on/off. Catalog restriction alone is not a treatment.

#### Scenario: Safe useful selector
- **WHEN** a candidate improves paired exact selection over the frozen base, produces zero destructive bypasses and zero out-of-scope action executions, and beats the incumbent risk/coverage frontier on a newly sealed set
- **THEN** it advances to deployed-quantized parity and resource measurement.

#### Scenario: Tiny capacity floor
- **WHEN** no arm passes the preregistered development gate across independent seeds
- **THEN** the 44.9M-scale path closes as a capacity-boundary result and the next candidate starts at 1.7B without further threshold search.

### Requirement: Complete learning receipt
Every stopped lane SHALL still publish the discriminating hypothesis, treatment, outcome, causal caveat, and next admissible experiment.

#### Scenario: Negative result
- **WHEN** a primary or regression gate fails
- **THEN** the ledger records the failed hypothesis without converting it into a capability win.

## Tasks

- [x] 1. Audit and reconcile WebGPU speed claims against retained raw evidence.
- [x] 2. Add structured experiment manifests and validators for the four conversion lanes.
- [x] 3. Upgrade the WebGPU runner to reject software adapters and emit paired JSON receipts; run the bounded hardware gate.
- [x] 4. Add a dependency-free ASR scorer and frozen reference fixture; run Parakeet and the available native baseline on identical audio.
- [x] 5. Freeze ReST baseline/candidate commands, frontier check, slices, and resource capture; download and run only after the no-training publish check passes.
- [x] 6. Build Needle training/dev data from independent sources, freeze sealed V2, run tiny-overfit and the 2x2 factorial, then test the selected quantized artifact once.
- [x] 7. Update the attempt ledger, report cards, public artifacts, recipes, learning path, and decision matrix from measured outcomes.
- [x] 8. Run focused tests, completion/quality/build gates, publish reviewed source, verify CI/deploy/live pages, and close this issue with receipts.

**Comment by sarthakagrawal927** [comment]
Task 1 audit receipt (local, not yet committed/published):

- The 2.6x-12.1x WebGPU curve entered the site in `ff7e903`, but no raw timing receipt or exact adapter identity survives.
- `browser/test_webgpu_train.mjs` proves 50-step final-loss parity; it does not emit timing data and accepts any adapter returned by `requestAdapter()`.
- `docs/performance/performance.md` is the later canonical correction: headless Chromium exposes SwiftShader and real-device performance is unknown until a hardware-qualified pair is recorded.
- Public speedup, roadmap, devlog, playground metadata, artifact registry, and attempt ledger are now corrected locally to label the curve historical/unqualified.
- Focused verification passes: attempt-ledger smoke (17/17), browser unit tests (79/79), TypeScript, production site/docs/agent build, and 109,894 generated internal links.
- ReST no-training preflight fails closed because no local artifact exists; it requires the pinned HF download after the eval contract is reviewed.
- The Parakeet receipt retains only the transcript hash, not the hypothesis text; the exact Kennedy sample therefore cannot be scored post hoc. Current upstream now publishes a v3 LibriSpeech route, which is a cleaner candidate for the fresh controlled accuracy experiment.

**Comment by sarthakagrawal927** [comment]
ASR qualification update (corrected after local binary check):

- The pinned v2 smoke cannot be scored retrospectively because only the transcript hash survives.
- Current upstream FluidInference/fluidaudio-web has moved to Parakeet v3 and now includes LibriSpeech/FLEURS WER runners; its README reports 2.15% LibriSpeech test-clean WER as upstream evidence. We must still reproduce on our frozen browser fixture before calling it a PostTrainLLM result.
- The fresh lane should therefore evaluate current v3 on an independently frozen public subset, retain hypotheses and references, and treat the old v2 run only as a runtime baseline.
- Both `/opt/homebrew/bin/whisperkit-cli` and `/opt/homebrew/bin/whisper-cli` are already available locally, as is ffmpeg. That removes the baseline-binary installation blocker; model/fixture availability still needs a read-only preflight before the controlled run.

Primary upstream: https://github.com/FluidInference/fluidaudio-web

**Comment by sarthakagrawal927** [comment]
Pre-implementation truth audit found and locally corrected additional WebGPU evidence leaks beyond the artifact page: README headline metrics, roadmap/social metadata and chart labels, devlog/RSS, playground runtime personalization, backend explainers, study guide, deployment guide, decision log, PLAN, and the runtime-boundary learning lab. All historical 2.6x-12.1x values now say that raw timings and adapter identity are missing.\n\nThe same pass found dead runtime explainer affordances: dynamically added shader-f16 and cooperative-matrix pills referenced missing explainer keys and were added after popover initialization. The local fix defines capability-vs-active explanations, initializes dynamic popovers, prevents duplicate cooperative-matrix pills, and rewrites f16/cooperative-matrix/WebNN copy so numerics gates and tiny probes are not presented as speed or full-model wins. A new parity test rejects any statically referenced data-explain key without content.\n\nLocal verification: 80/80 browser tests, TypeScript, attempt-ledger 17/17, project-completion audit (75 experiments / 18 recipes / 9 paths / 13 buildable artifacts / 17 public artifacts / 0 unresolved), full site+docs+agent build (426 HTML pages, 358 Markdown counterparts, 109,895 internal links). These changes are not committed or published pending authorization.

**Comment by sarthakagrawal927** [comment]
Owner review cleared on 2026-09-02 with: “What are the blockers? Just keep pushing.” Proceeding with the frozen #138 protocol, separate reviewable commits, and the bounded resource envelopes already specified. Apple notarization remains a separate credential-dependent release gate.

**Comment by sarthakagrawal927** [comment]
Task 2 complete and pushed in `3aabf83`.

Receipts:
- Four design-frozen manifests: WebGPU, Parakeet/native ASR, unchanged-weight ReST requalification, and the full 2x2 Needle successor.
- `scripts/experiments/validate_verified_win_manifest.py` enforces shared evidence fields plus lane-specific gates.
- `--stage run` fails closed until revisions, fixture hashes, executable commands, and open requirements are resolved.
- `evals/verified-win-manifest-smoke.sh` passes and proves incomplete Needle designs and design-only run attempts are rejected.
- Format and lint/typecheck pass.

**Comment by sarthakagrawal927** [comment]
Task 3 complete. Published result: `2a01a5a`.

Measured receipt (`evals/verified-wins/webgpu-paired-result-v1.json`):
- Apple M5 Pro, Chrome 151, adapter `apple · metal-3`; fallback=false, software=false.
- Frozen Large preset, TinyShakespeare, seed 42, 20 steps/arm, ABBA order.
- WASM median: 1462.5 ms/step.
- WebGPU median: 137.1 ms/step.
- Median speedup: **10.67x** (gate >1.25x).
- Maximum paired final-loss drift: **4.72%** (gate <5%).
- Runtime errors: 0.
- Decision: **promote** for this exact machine/configuration; other preset points remain historical.

Gates passed: manifest run validation, fail-closed smoke, attempt-ledger smoke, 81 browser tests, format, lint/typecheck, completion checks, production build, and built-link audit.

**Comment by sarthakagrawal927** [comment]
Task 4 receipt: completed and pushed in `157a2aa` (frozen runners) and `15fd008` (measured result + public learning artifact). On the same eight LibriSpeech test-clean clips (34.365s, 82 words), browser Parakeet v3 scored 0/82 word errors versus native WhisperKit large-v3 turbo at 7/82 (8.54% WER), matched 4/4 proper nouns, produced 0 repetition events, made 0 external warm requests, and decoded 3.51x faster overall (914.7ms vs 3211.8ms). Real Apple/Metal 3 WebGPU was verified; the browser downloaded 684,385,250 pinned bytes. The preregistered 50x median short-clip RTF gate failed at 33.84x, so the formal decision remains `reject` / `worked-with-caveat` rather than moving the threshold. Tracked result: `evals/verified-wins/parakeet-asr-result-v1.json`; protocol: `evals/verified-wins/parakeet-asr-v1.json`; raw files are hash-addressed in the tracked result.

**Comment by sarthakagrawal927** [comment]
Tasks 5–7 are complete and published through `ac9c6d3`.

ReST requalification:
- Frontier ceiling passed: depth 12/12, breadth 44/45.
- Stock → ReST depth: 9/12 → 12/12; unexpected side effects 8 → 0; depth wall time 360.50s → 148.91s (2.42×).
- Stock → ReST breadth: 30/45 → 25/45 (-11.1pp); paired stock-only 9 vs candidate-only 4.
- Decision: retain only behind the file-operations route; reject as the general successor.
- Result: `evals/verified-wins/rest-requalification-result-v1.json`.

Needle successor:
- All four 45M arms passed exact tiny overfit.
- Public-development exactness: 24/94, 24/94, 25/94, 26/94 vs stock 32/94.
- Every arm produced all 10 destructive bypasses vs stock 2, triggering the predeclared unsafe stop before sealed V2.
- Decision: `advance-model-class`; close 45M recipe search. A 1.7B continuation requires a fresh scoped issue.
- Result: `evals/verified-wins/needle-successor-result-v1.json`; complete hands-on recipe: `docs/techniques/needle2-successor-factorial.md`.

Accounting and learning surfaces:
- 76/76 experiments have final dispositions: 5 worked, 37 worked-with-caveat, 34 non-positive/mixed.
- 18 recipe contracts, 9 learning paths, 13 buildable journey artifacts, and 17 public artifacts pass the completion validator.
- ReST/Needle artifacts, package/model cards, homepage proof, report-card cohort, registries, and agent surfaces now use the fresh evidence.
- Hero loss curve has a 56px enforced clearance above the metric row at 390/768/1440.

Local gates: 81 browser tests; attempt-ledger, project-completion, fine-tune-report-card, quality, production build, 359 human/agent page pairs, 110,565 internal links, and 36 responsive route/viewport checks all pass. Source is pushed at `ac9c6d3`; current-SHA CI run 33752510298 is in progress. Task 8 remains open until CI completes. Deployment/live parity and Apple notarization are separate authorization/credential gates and are not being claimed.

**Comment by sarthakagrawal927** [comment]
Current-SHA CI is green: run 33752510298 passed all eight jobs on `ac9c6d3705985432c5067226f9d51756634d6bfd` (macOS build/Xcode tests, eval smokes, Swift format, gallery drift, WASM drift, Python, browser, and code health).

Task 8 is complete through reviewed source, push, and exact-SHA CI. It remains intentionally unchecked because its final clause requires deployment and live-page verification; no deployment authorization is being inferred from the experiment approval. The four experiments themselves, their decisions, and their learning artifacts are complete.

**Comment by sarthakagrawal927** [comment]
Completed and live.

Receipts:
- Source: `ac9c6d3705985432c5067226f9d51756634d6bfd` on `main`, clean and synced.
- Current-SHA CI: https://github.com/PostTrainLLM/posttrainllm/actions/runs/33752510298 — all jobs passed.
- Production deploy: https://github.com/PostTrainLLM/posttrainllm/actions/runs/33754199645 — build, Cloudflare Pages upload, and production smoke passed.
- Live machine catalog: https://posttrainllm.com/api/ai — 76/76 resolved; 5 worked, 37 worked-with-caveat, 34 non-positive/mixed; 9 paths, 18 recipes, 13 buildable artifacts.
- Live routes verified `200`: homepage, experiments, artifacts, learning paths, recipes, Web Lab, Mac download, `llms.txt`, `llms-full.txt`, `/api/ai`, and OpenAPI.
- Live desktop visual check confirms the hero curve stays above the metric row without the masking panel.

The verified-win tranche is complete. Apple notarization remains separately tracked by #137.
